### PR TITLE
Run cross version tests when we update maximum versions

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

## What changes are proposed in this pull request?

When we update maximum versions for making a new release, run tests for udpated packages.

## How is this patch tested?

Existing check

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: dev/set_matrix.py [hunk 1]
```diff
@@ -419,6 +419,7 @@ def expand_config(config):
                         run=run,
                         package=pip_release,
                         version=ver,
+                        newer_than_max=Version(ver) > Version(max_ver),
```
Comment (harupy) [inline dev/set_matrix.py:422]: We update `ml-package-versions.yml` when we make a release like below:

```diff
# ml-package-versions.yml

sklearn:

  autologging:
    minimum: "0.20.3"
-    maximum: "0.24.1"
+    maximum: "0.24.2"
```

For this change, we want to run tests for sklearn autologging using sklearn 0.24.2 to ensure sklearn autologging really works with 0.24.2. However, `set_matrix.py` skips running tests now. Here' why:


```python
# Assumption: the latest version of sklearn is 024.2 now
# NOTE: our cross-version-tests strategy fetches the latest version available ignoring the maximum version specified in `ml-package-versions.yml`.

# ---------- NOW ----------
# Matrix generated from master
[
    {"package": "sklearn", "version": "0.24.2"},
]

# Matrix generated from PR branch
[
    {"package": "sklearn", "version": "0.24.2"},
    # -> considered as not-updated
]

# ---------- PROPOSED ----------
# Matrix generated from master (where sklearn maximum version is 0.24.1
# so `newer_than_maximum` evaluates to True)
[
    {"package": "sklearn", "version": "0.24.2", "newer_than_maximum": True},
]

# Matrix generated from PR branch (where sklearn maximum version is 0.24.2
# so `newer_than_maximum` evaluates to False)
[
    {"package": "sklearn", "version": "0.24.2", "newer_than_maximum": False},
    # -> considered as "udpated"
]

# `newer_than_maximum` field is different -> run test for sklearn 0.2

```

File: mlflow/ml-package-versions.yml [hunk 2]
```diff
@@ -13,7 +13,7 @@ sklearn:
 
   autologging:
     minimum: "0.20.3"
-    maximum: "0.24.1"
+    maximum: "0.24.2"
```
Comment (harupy) [inline mlflow/ml-package-versions.yml:16]: A dummy change to make sure the proposed change works fine.

Comment (harupy) [reply mlflow/ml-package-versions.yml:16]: It worked:

image

File: dev/set_matrix.py (hunk 1 above)
Comment (harupy) [reply dev/set_matrix.py:422]: This works but it's not straight forward.

Comment (WeichenXu123) [reply dev/set_matrix.py:422]: The code logic is a little strange.

Why not do:
```
# Matrix generated from master 
matrix_master = [ {"package": "sklearn", "max_supported_version": "0.24.1"}, ]

Matrix generated from PR branch:
matrix_pr = [ {"package": "sklearn", "max_supported_version": "0.24.2"}, ]
```

Then we get master and PR's max_supported_version, and then we can generate a series of CI runs. We don't need the field `newer_than_maximum` which make the logic hard to understand.

Comment (harupy) [reply dev/set_matrix.py:422]: If we did that, sklearn `0.24.2` would be always tested until  we update the max version to `0.24.2`.

Comment (harupy) [reply dev/set_matrix.py:422]: How about renaming `newer_than_max` to `supported`?

File: dev/set_matrix.py [hunk 3]
```diff
@@ -419,6 +419,7 @@ def expand_config(config):
                         run=run,
                         package=pip_release,
                         version=ver,
+                        supported=Version(ver) <= Version(max_ver),
```
Comment (harupy) [inline dev/set_matrix.py:422]: @WeichenXu123 I renamed `newer_than_maximum` to `supported` (which I thin is clearer).

File: mlflow/ml-package-versions.yml [hunk 4]
```diff
@@ -13,7 +13,7 @@ sklearn:
 
   autologging:
     minimum: "0.20.3"
-    maximum: "0.24.2"
+    maximum: "0.24.3"
```
Comment (harupy) [inline mlflow/ml-package-versions.yml:16]: This change should create a run for sklearn 0.24.3.

File: dev/set_matrix.py (hunk 3 above)
Comment (harupy) [inline dev/set_matrix.py:422]: When we update a package version, this field flips to `True` and the cross-version-tests script recognizes it as a new item. We could use this field to allow failure versions that we don't support by checking this field.

File: dev/set_matrix.py [hunk 5]
```diff
@@ -439,6 +440,7 @@ def expand_config(config):
                         run=run,
                         package=pip_release,
                         version=DEV_VERSION,
+                        supported=False,
```
Comment (harupy) [inline dev/set_matrix.py:443]: The dev version should always be unsupported.

File: dev/set_matrix.py [hunk 6]
```diff
@@ -473,7 +475,10 @@ def main():
     diff_flavor = set(filter(lambda x: x["flavor"] in changed_flavors, matrix))
 
     # If this file contains changes, re-run all the tests, otherwise re-run the affected tests.
-    include = matrix if (__file__ in changed_files) else diff_config.union(diff_flavor)
+    # I'LL REVERT THIS CHANGE BEFORE MERGE
+    # include = matrix if (__file__ in changed_files) else diff_config.union(diff_flavor)
+    include = diff_config.union(diff_flavor)
+    # I'LL REVERT THIS CHANGE BEFORE MERGE
```
Comment (harupy) [inline dev/set_matrix.py:481]: A temporary change to avoid running too many cross-version tests.

**Review by WeichenXu123 — APPROVED** [review]

File: mlflow/ml-package-versions.yml [hunk 7]
```diff
@@ -236,9 +236,9 @@ fastai-1.x:
       pytest tests/fastai/test_fastai_model_export.py --large
 
   autologging:
-    pin_maximum: True
+    # pin_maximum: True
     minimum: "1.0.60"
-    maximum: "1.0.61"
+    maximum: "2.0.0"
```
[the remark this answers is not in the document]
Comment (WeichenXu123) [reply mlflow/ml-package-versions.yml:241]: Why CI triggered 2.1/2.2/2.3 version tests ? shouldn't it only trigger 2.0 version test ?

[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/ml-package-versions.yml:241]: Yep, this is a temporary change to make sure the proposed logic works. The maximum version actually specifies the maximum **major** version. For example, `maximum: "2.0.0"` basically means `test all 2.x minor release but`.

**Pull request merged**